subprojects/core/src/main/java/org/gradle/api/internal/DefaultNamedDomainObjectCollection.java

```diff
@@ -144,7 +145,20 @@ public boolean addAll(Collection<? extends T> c) {
     }
 
     @Override
-    public void addLater(final Provider<? extends T> provider) {
+    public void addLater(Provider<? extends T> provider) {
+        assertMutable("addLater(Provider)");
+        if (provider instanceof Named) {
+            final Named named = (Named) provider;
+            String name = named.getName();
+            if (index.get(name) != null || index.getPending(name) != null) {
```

Comment:
If coverage for this case (registering a provider with the same name as an existing provider or realized element) is missing, we should add some in the unit tests.

Reply:
I added test coverage for this.